ClassUtil tests + bugfix

er1c opened the pull request:
Added some `ClassUtil` tests and found a couple recursion bugs in the process.

er1c commented:
Updated to include fixes that will make tests pass on win32.  The `lastModified` tests are a bit borked.  @tpunder should those be based upon a macro?  Tempted to just comment them out for now

er1c commented on jvm/src/main/scala/fm/common/ClassUtil.scala:
```diff
@@ -212,16 +212,17 @@ object ClassUtil extends Logging {
+        case "jar" =>
+          val jarFile: String = filePath.substring("file:".length, filePath.indexOf("!"))
+          val jarPrefix: String = filePath.substring(filePath.indexOf("!") + 1)
+          require(jarPrefix == s"/$packageDirPath", s"Expected jarPrefix ($jarPrefix) to equal package prefix (/$packageDirPath)")
```
I probably don't have a good test for this code change, now that I think about it.  On the plus side, hard coding for "/" won't affect the linux/OSX dev @tpunder do you know what this is suppose to cover?

tpunder replied:
The URLDecoding behavior is now different so this needs test for sure to verify that the existing behavior is broken and that the new behavior is correct.

tpunder replied:
Also, using interpolation here seems like overkill and I prefer the original style: `File.separator+packageDirPath` and `packageDirPath+File.separator`

er1c replied:
The `jar` test remains the same, I think just the file-test is different.  I'll see about adding an explicit test for this.  I think I ran into this because on win32 I got back a  url.getFile() like: `c:/Users/Eric%20Peters/IdeaProject/fm-common/...` and it didn't work unless I url decoded the path before creating a `new File(path)` out of it.

tpunder commented:
Here is a local change I had to partially fix the classpath separator issue.  Note how I added a `classpathSeparator` method that could be referenced in place of `File.separator`.  I think this is better than just referencing `/` everywhere.

```diff
$ git show 81ed2b51a3a6fc051ac836a9d0b94e32d13a90c8
commit 81ed2b51a3a6fc051ac836a9d0b94e32d13a90c8
Author: Tim Underwood <timunderwood@gmail.com>
Date:   Thu Aug 3 11:25:52 2017 -0700

    Fixes https://github.com/frugalmechanic/fm-common/issues/18

diff --git a/jvm/src/main/scala/fm/common/ClassUtil.scala b/jvm/src/main/scala/fm/common/ClassUtil.scala
index a0e4058..fb84b13 100644
--- a/jvm/src/main/scala/fm/common/ClassUtil.scala
+++ b/jvm/src/main/scala/fm/common/ClassUtil.scala
@@ -259,7 +259,13 @@ object ClassUtil extends Logging {
     builder.result
   }

-  private def getPackageDirPath(basePackage: String): String = basePackage.stripLeading(File.separator).replace(".", File.separator)
+  // Note: The classpath separator is *ALWAYS* a / and should not be File.separator
+  // See:
+  //   https://www.atlassian.com/blog/archives/how_to_use_file_separator_when
+  //   http://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#getResource-java.lang.String-
+  private def classpathSeparator: String = "/"
+
+  private def getPackageDirPath(basePackage: String): String = basePackage.stripLeading(classpathSeparator).replace(".", classpathSeparator)

   private def defaultClassLoader: ClassLoader = {
     val cl: ClassLoader = Thread.currentThread.getContextClassLoader
```

tpunder commented on jvm/src/test/scala/fm/common/TestClassUtil.scala:
```diff
@@ -0,0 +1,153 @@
+      updatedTestClasses.map{ _.replace(".", "/") + ".class" } ++ Set("test/classutil/lorem-ipsum.txt", "test/classutil/subpackage", "test/classutil/subdirectory")
+    }.map{ new File(_) }
+
+    ClassUtil.listClasspathFiles("test.classutil") should equal(expectedFiles)
```
Can you add tests for `ClassUtil.listClasspathFiles("")` and `ClassUtil.listClasspathFiles("/")` ?

tpunder commented on jvm/src/test/scala/fm/common/TestClassUtil.scala:
```diff
@@ -0,0 +1,153 @@
+      testClasses.map{ _.replace(".", "/") + ".class" } ++ Set("test/classutil/lorem-ipsum.txt", "test/classutil/subdirectory/subfile.txt")
+    }.map{ new File(_) }
+
+    ClassUtil.findClasspathFiles("test.classutil") should equal (expectedFiles)
```
Can you add tests for `ClassUtil.findClasspathFiles("")` and `ClassUtil.findClasspathFiles("/")` ?

er1c commented:
PR Updated, I added the requested changes/new test scenarios (those were suppose to return exceptions, right?)  I like that `private def classpathSeparator: String = "/"` convention, it makes the code a bit easier to read than the `"/"` floating around.

I also added tests for `findAnnotatedClasses` and `findImplementingClasses`

Discovered that the findAnnotatedClasses doesn't work with Scala annotations at all...but I didn't see an obvious way to fix that.  I added the test classes, but just left them commented out.

er1c commented:
Ooops I missed the `findClasspathFiles("")` tests

tpunder commented on jvm/src/test/scala/fm/common/TestClassUtil.scala:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  test("""listClasspathFiles("/")""") {
+    assertThrows[IllegalArgumentException] {
```
I don't think this is the intended behavior

tpunder commented on jvm/src/test/scala/fm/common/TestClassUtil.scala:
```diff
@@ -0,0 +1,163 @@
+
+  test("""listClasspathFiles("")""") {
+    assertThrows[IllegalArgumentException] {
+      ClassUtil.listClasspathFiles("")
```
I'm not sure if this is the intended behavior.  Do we require at least a `/` for things to work?

er1c replied:
I'm sorta looking into it...I think the implementation is broken for the `listClasspathFiles`, because the `findClasspathFiles` does "work" in one of these situations.

tpunder commented on jvm/src/main/scala/fm/common/ClassUtil.scala:
```diff
@@ -212,16 +218,17 @@ object ClassUtil extends Logging {
-          val jarPrefix: String = fullJarUrl.substring(fullJarUrl.indexOf("!") + 1)
-          require(jarPrefix == File.separator+packageDirPath, s"Expected jarPrefix ($jarPrefix) to equal package prefix ($packageDirPath)")
-          scanJar(packageDirPath+File.separator, new File(jarFile))
+        case "jar" =>
```
Do any of the tests cover the `"jar"` code path?

er1c replied:
They currently *don't*, but I didn't change the code behavior for the `"jar"` path, i just factored out the `val fullJarUrl: String = URLDecoder.decode(url.getFile(), "UTF-8")` into  `val filePath: String = URLDecoder.decode(url.getFile(), "UTF-8")` above.  I'll look into figuring out if there is a good test to do here based upon our current dependencies.

er1c commented:
PR Updated (again), but added `"jar"` tests (which I'm using the `fastutil` library for)  I also fixed the `listClasspathFiles` bug that was uncovered by adding `listClasspathFiles("/")` and `listClasspathFiles("")` tests.

tpunder commented on jvm/src/test/scala/fm/common/TestClassUtil.scala:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  test("findClassNames - defaultClassLoader - jar file") {
+    ClassUtil.findClassNames("it.unimi.dsi") should contain ("it.unimi.dsi.fastutil.ints.IntOpenHashSet")
```
At some point I plan to split this out into an `eluvio-common` project that has no external dependencies.  So can you choose something else to test here?  Testing Scala standard library stuff is probably okay.

er1c commented:
PR Updated, now using `scala.collection` for the `"jar"` tests.

er1c commented:
Confirmed that this branch's test successfully on win32:

```
[info] Run completed in 5 seconds, 311 milliseconds.
[info] Total number of tests run: 162
[info] Suites: completed 31, aborted 0
[info] Tests: succeeded 162, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 25 s, completed Aug 9, 2017 3:47:31 PM
```

er1c commented:
I can also verify the `ta-catalog` `./sbtTestWebservice.sh` works with this change, without any other code-changes in `ta-catalog` for `win32`

tpunder commented:
Released as `0.10.0`
